ruby/sdk-phase-1.md

```diff
@@ -0,0 +1,439 @@
+# Ruby - Phase 1
+
+The Ruby SDK will be implemented in two loosely defined phases:
+
+* Phase 1 - Initial project, activity worker, client, and test framework
+* Phase 2 - Workflow worker and replayer
+
+This proposal is for phase 1. While proposals and SDK may be done in large phases, work is actually incremental.
+
+Notes for this proposal:
+
+* Everything is subject to change both during proposal time and afterwards during implementation.
+* For each place a decision is made, a "💭 Why X? ..." aside is present explaining why.
+* For each place an open question exists, a "❓ X?" exists to ask it.
+
+## Overview
+
+The Ruby SDK will be backed by Rust [sdk-core](https://github.com/temporalio/sdk-core) and will follow the model set by
+other SDKs.
+
+Much of the work for this proposal and the implementation comes the past proposals in [previous](previous)
+and the implementation built from that. 💭 Why start a new proposal/implementation instead of adapting that? The
+constraints, approach, and requirements changed enough to justify a clean slate and while many things may be leveraged,
+it never went beyond alpha anyways.
+
+Lots of approaches taken here are similar to recent Python and .NET SDKs.
+
+### Requirements
+
+These are the general requirements of the project
+
+* Ruby >= 3.1
+  * 💭 Why not older versions? They are all EOL'd and we need Ractors and other things that come with 3.x.
+* Limited dependencies
+  * This means we will try to make `google-protobuf` the only dependency. We should not depend on `async` or anything
+    else if we can help it.
+  * 💭 Why? Our dependencies as a library have transitive effects on our users, so we have a responsibility to be
+    lightweight.
+* Must be both thread-blocking/thread-safe and async-capable
```

Comment:
What do you mean exactly by "async-capable"? (The term is used in a few places)

Reply:
In Ruby there are threads and fibers (basically coroutines). By async-capable I mean fiber-capable. So you can use https://github.com/socketry/async and, say, use the client to wait on 100 workflow results without using a thread each. But many Ruby users are totally fine using a thread each and we need to support that too.

This is very much like Python, we just chose not to provide a synchronous version of the Python client (though we still could if we wanted).